### PR TITLE
Quarterdeck passthrough credentials

### PR DESCRIPTION
**Pull request by pdeziel**
### Scope of changes

This modifies the quarterdeck client and middleware to facilitate passthrough credentials. This approach uses the context that's passed through the client handlers and on the middleware side, the access token is initially passed through the gin context on `Authenticate` so it's available when the user of the quarterdeck client creates the request context.

Fixes SC-12797

### Type of change

- [x] new feature
- [ ] bug fix
- [ ] documentation
- [ ] testing
- [ ] technical debt
- [ ] other (describe)

### Acceptance criteria

Implementors who are using the quarterdeck client (e.g. the tenant handlers) should be able to use the `ContextFromRequest` helper to create the request context with the authentication token in it in order to facilitate the passthrough.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [x]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.

### CODE REVIEW

File: pkg/quarterdeck/middleware/auth.go [hunk 1]
```diff
@@ -128,6 +130,24 @@ func GetClaims(c *gin.Context) (*tokens.Claims, error) {
 	return claims.(*tokens.Claims), nil
 }
 
+// ContextFromRequest creates a context from the gin request context, copying fields
+// that may be required for forwarded requests. This method should be called by
+// handlers which need to forward requests to other services and need to preserve data
+// from the original request such as the user's credentials.
+func ContextFromRequest(c *gin.Context) (ctx context.Context, err error) {
```
Comment (pdeziel) [inline pkg/quarterdeck/middleware/auth.go:137]: Note that I didn't end up creating an interface implementation of the passthrough credentials and used the `Token` one instead, let me know what you think.

File: pkg/quarterdeck/api/v1/client_test.go [hunk 2]
```diff
@@ -78,6 +78,34 @@ func TestClient(t *testing.T) {
...
+	require.True(t, ok, "could not cast client to APIv1")
+	req, err = apiv1.NewRequest(context.Background(), http.MethodPost, "/bar", data, nil)
+	require.NoError(t, err, "could not create request")
+	require.Equal(t, "Bearer default", req.Header.Get("Authorization"), "expected the authorization header to be set to default")
+
+	// Test that request credentials override default credentials
+	ctx = api.ContextWithToken(context.Background(), "newtoken")
+	req, err = apiv1.NewRequest(ctx, http.MethodPost, "/bar", data, nil)
+	require.NoError(t, err, "could not create request")
+	require.Equal(t, "Bearer newtoken", req.Header.Get("Authorization"), "expected the authorization header to be set to newtoken")
```
Comment (bbengfort) [inline pkg/quarterdeck/api/v1/client_test.go:108]: Very good set of tests!

File: pkg/quarterdeck/middleware/auth.go [hunk 3]
```diff
@@ -128,6 +130,24 @@ func GetClaims(c *gin.Context) (*tokens.Claims, error) {
...
+	var req *http.Request
+	if req = c.Request; req == nil {
+		return nil, ErrNoRequest
```
Comment (bbengfort) [inline pkg/quarterdeck/middleware/auth.go:140]: I would be very surprised if this ever happened … I regularly use `c.Request.Context()` without doing `nil` check on `c.Request` -- but this is very thorough!

File: pkg/quarterdeck/middleware/auth.go (hunk 1 above)
Comment (bbengfort) [reply pkg/quarterdeck/middleware/auth.go:137]: I'm willing to give this a shot to see how we like using it!

**Review by bbengfort — APPROVED** [review]
Looks good! I'm going to merge this in so that it unblocks the work I'm doing on the Quarterdeck API. I will likely do a little reorganization and clean up the old `RequestOptions` things in a following PR - so please make sure you check the state of things and do a pull from `main` before you get started on Monday!

**Pull request merged**